zion/src/main/scala/edu/uci/ics/cloudberry/zion/model/impl/SparkSQLGenerator.scala

```diff
@@ -0,0 +1,652 @@
+    val producedExprs = mutable.LinkedHashMap.newBuilder[String, FieldExpr]
+    val unnestTestStrs = new ListBuffer[String]
+    val unnestStr = unnest.zipWithIndex.map {
+      case (unnest, id) =>
+        val expr = exprMap(unnest.field.name)
+        val newExpr = s"${quote}hashtag$quote"
+        producedExprs += (unnest.as.name -> FieldExpr(newExpr, newExpr))
+        if (unnest.field.isOptional) {
+          unnestTestStrs += s"${expr.refExpr} is not null"
+        }
+        s"lateral view explode(${expr.refExpr}) tab as $newExpr"
+    }.mkString("\n")
+    appendIfNotEmpty(queryBuilder, unnestStr)
+
+    ParsedResult(unnestTestStrs.toSeq, (producedExprs ++= exprMap).result().toMap)
+  }
+
+  protected def parseGroupByFunc(groupBy: ByStatement, fieldExpr: String): String = {
+    groupBy.funcOpt match {
+      case Some(func) =>
+        func match {
+          //          case bin: Bin => s"${round}($fieldExpr/${bin.scale})*${bin.scale}"
+          //          case interval: Interval =>
+          //            val duration = parseIntervalDuration(interval)
+          //            s"${typeImpl.getIntervalStartDatetime}(${typeImpl.intervalBin}($fieldExpr, '1990-01-01T00:00:00.000Z', $duration))"
+          case interval: Interval => s"$fieldExpr"
+          //get_interval_start_datetime(interval_bin(t.`create_at`, '1990-01-01T00:00:00.000Z',  day_time_duration("PT1H") ))
+          case level: Level =>
+            //TODO remove this data type
+            val hierarchyField = groupBy.field.asInstanceOf[HierarchyField]
+            val field = hierarchyField.levels.find(_._1 == level.levelTag).get
+            s"$fieldExpr.${field._2}"
+          //          case GeoCellTenth => parseGeoCell(10, fieldExpr, groupBy.field.dataType)
+          //          case GeoCellHundredth => parseGeoCell(100, fieldExpr, groupBy.field.dataType)
+          //          case GeoCellThousandth => parseGeoCell(1000, fieldExpr, groupBy.field.dataType)
+          case _ => throw new QueryParsingException(s"unknown function: ${func.name}")
+        }
+      case None => fieldExpr
+    }
+  }
+
+  private def parseGroupby(groupOpt: Option[GroupStatement],
+                           exprMap: Map[String, FieldExpr],
+                           queryBuilder: StringBuilder): ParsedResult = {
+    groupOpt match {
+      case Some(group) =>
+        val producedExprs = mutable.LinkedHashMap.newBuilder[String, FieldExpr]
+        val groupStrs = group.bys.map { by =>
+          //          println("exprMap: ", exprMap)
+          val fieldExpr = exprMap(by.field.name)
+          val as = by.as.getOrElse(by.field)
+          val groupExpr = parseGroupByFunc(by, fieldExpr.refExpr)
+          val newExpr = s"$quote${as.name}$quote"
+          producedExprs += (as.name -> FieldExpr(newExpr, newExpr))
+          //          println("newExpr: ", newExpr)
+          if (newExpr == s"${quote}hashtags${quote}" || newExpr == s"${quote}tag${quote}"){
```

Comment:
check if query is grouped by hashtags or state. If it is, then no group by function will be added, it will be "group by `hashtag` or group by `state`" directly. Otherwise, a "group by" function will be used: for instance, "group by month(t.`create_at`)"